src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1419,6 +1433,150 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
     setFileStats();
 }
 
+bool
+DatabaseShardImp::callForLedgerSQL(
+    LedgerIndex ledgerSeq,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+    auto shardIndex = seqToShardIndex(ledgerSeq);
+
+    if (shards_.count(shardIndex) && shards_[shardIndex].state == ShardInfo::State::finalize)
+    {
+        return shards_[shardIndex].shard->callForLedgerSQL(callback);
+    }
+
+    return false;
+}
+
+bool
+DatabaseShardImp::callForTransactionSQL(
+    LedgerIndex ledgerSeq,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+    auto shardIndex = seqToShardIndex(ledgerSeq);
+
+    if (shards_.count(shardIndex) && shards_[shardIndex].state == ShardInfo::State::finalize)
+    {
+        return shards_[shardIndex].shard->callForTransactionSQL(callback);
+    }
+
+    return false;
+}
+
+bool
+DatabaseShardImp::iterateLedgerSQLsForward(
+    LedgerIndex minShardIndex,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+
+    std::map<std::uint32_t, ShardInfo>::iterator it, eit;
+
+    if (minShardIndex == (LedgerIndex)-1)
+        it = shards_.begin();
+    else
+        it = shards_.lower_bound(minShardIndex);
```

Comment:
So the first returned shard will be strictly greater than `minShardIndex`? Is there a reason it can't be greater than or equal? (if it can be `>=` then we can use `0` for the special value instead of the proposed optional). At any rate, we definately need to document that `minShardIndex` means strictly greater, as I would have expected `>=` from the param name.

Reply:
lower_bound is >=, upper is >; fixed code with upper_bound

Reply:
I don't see this change. However, does this mean that you meant `upper_bound` in the original code and this was a bug? Or do we need to change the callers as well to account for the fact we changed the semantics here? (also, if we use `upper_bound` we can remove the `optional` param and just pass in 0).

Reply:
What change do you mean?

> So the first returned shard will be strictly greater than `minShardIndex`?

It seems not true. lower_bound(x) returns x if x exists. See C++ manuals please, [for ex.](http://www.cplusplus.com/reference/map/map/lower_bound/)

Reply:
You're right, I was confused (apologies). However, now I don't understand why `minShardIindex` needs to be an `optional`. Can't we just pass in `0` and get the same behaviour as passing in `std::nullopt`?